How to create an ID token with a groups claim

mystiberry-okta opened the pull request:
## Description:
- How to create an ID token with a groups claim

### Resolves:
 [OKTA-134219](https://oktainc.atlassian.net/browse/OKTA-134219)

mystiberry-okta commented:
Here's a PDF of the new topic. 
[groups-claim.pdf](https://github.com/okta/okta.github.io/files/1209802/groups-claim.pdf)

mystiberry-okta commented:
@gronberg First draft ready for review. I didn't include the results, or a screen shot of the decoded JWT, because I couldn't get the /token request to work yet. But let me know what you think of this overall. And if anyone else should review it.

johngronberg-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+
+To do this, create an OpenID Connect client, assign a group whitelist to it, and configure a groups claim: 
+
+1. Create an OpenID client at `/{yourOktaDomain}.com/api/v1/apps`
```
Maybe we add a pre-requisite that you have an OIDC client, rather than telling the reader to make a new one.

johngronberg-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+            } 
+    ~~~
+    
+3. PUT the group IDs in the profile property bag: /{yourOktaDomain}.com/api/v1/apps/:clientId
```
This is an optional step. The reader could just add the one id to the actual Okta EL function. If they want to dynamically set group whitelists on a per application basis, then they can use the profile property bag. We added the property bag step to help customers who have many apps that care about different groups, so they wouldn't have to whitelist 1000+ groups in the original EL function.

cwu-okta replied:
PUT the group IDs in the profile property bag: -> Gather the group IDs and PUT them into the app's profile property bag:

mystiberry-okta commented:
From hipchat: makes sense to mention you can do this with either OIDC or API AM (include min. # of screenshots for clarity).

cwu-okta commented:
This PR is very hard to review as it contains changes unrelated to the asked for guide. Can you rebase and clean the PR up so that it only contians the related commit(s)?

mystiberry-okta commented:
Cleaned up!

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+
+## How to Create a Token with an Apps Groups Claim
+
+Because Okta allows you to specify a profile with whatever JSON-compliant contents you wish, you can create a list of groups that are included in the token
```
Nit: The tone of the first sentence is a little loose. Could we make it a bit more formal? Maybe something along the lines of "Okta provides a flexible profile which accepts any JSON-compliant content."

mystiberry-okta replied:
done.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+## How to Create a Token with an Apps Groups Claim
+
+Because Okta allows you to specify a profile with whatever JSON-compliant contents you wish, you can create a list of groups that are included in the token
+in the groups claim. You can add a group claim for any app (Okta, Active Directory, or Workday, for example) into ID tokens or access tokens for API Access Management.
```
Okta is not an app. It would be better to omit Okta from the list of example 'apps'.

Since you are trying to convey User Groups can also be added to claims I suggest rewording to 
```You can add a group claim for any app (Active Directory, or Workday, for example) or user group into ID tokens or access tokens for API Access Management.```

mystiberry-okta replied:
done. I just removed the examples as they made the sentence hard to read. Shouldn't have to explain what an app is at this point.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+   Request Example:
+    ~~~
+    curl -X GET \
+      http://mysticorp.oktapreview.com/api/v1/groups \
```
mysticorp -> yourOktaDomain
oktapreview -> {removed}

mystiberry-okta replied:
done. also fixed the session token again.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+      -H 'content-type: application/json' \
+      -H 'postman-token: 4717a190-ebf9-d68c-f334-12458ccd1a4c' \
+      -d '{
+      "profile": {
```
Why is there a profile in the GET request?

mystiberry-okta replied:
Removed.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+            ],
+            "type": "BUILT_IN",
+            "profile": {
+                "name": "Everyone",
```
The Everyone group is a very special group. I suggest updating the example with a 'normal' group

mystiberry-okta replied:
Fixed.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+    
+3. PUT the group IDs in the profile property bag: /{yourOktaDomain}.com/api/v1/apps/:clientId
+    
+    Request Example Fragment:
```
I also ask why the request is a fragment and not the whole request? For the PUT you need to have the entire object (not a partial one.) Users might (mistakenly) just read too quickly and try to make the request according to teh guide and receive back an error since they did not put the whole object. Normally we say to use the response from the Step x and amend it to contain the profile object as shown.

mystiberry-okta replied:
Fixed.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+       },
+    ~~~
+    
+    You can select any group or set of groups, either app groups or user groups. 
```
Because you said group or app group here, I would make sure the above example contains either one or the other, but both would be great too :)

```You can select any group or set of groups, either app groups or user groups.``` The 'any group or set of groups' is awkward. Maybe reword to ```You can add either app groups or user groups to the group white list property. The property value must be an array.```

mystiberry-okta replied:
done.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+
+    If you want to use this group whitelist for every client that gets this claim in a token, put the `groupId` in the first parameter of the `getFilteredGroups` function described below. 
+ 
+4. Add an ID token claim on Custom Authorization Server in the Okta UI with the following function: `getFilteredGroups({group ID list}, {value-to-represent-the-group-in-the-token}, {maximum number of groups to include in token. Must be less than 100.})`.
```
Why does the entire guide talk about using the API then switch to using the UI to accomplish some work? Why not stick with just the API? If you talk about the UI here it makes it seem like our API doesn't support this, which it does.

mystiberry-okta replied:
done.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+
+    ~~~curl
+    curl -X POST \
+      http://mysticorp.oktapreview.com/oauth2/ausain6z9zIedDCxB0h7/v1/token \
```
mysticorp -> myOktaDomain

mystiberry-okta replied:
Done.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+excerpt: How to create an ID token or access token which contains a groups claim
+---
+
+## How to Create a Token with an Apps Groups Claim
```
How to Create a Token with an Apps Groups Claim -> How to Create a Token with a Groups Claim

mystiberry-okta replied:
done.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+    Request Example Fragment:
+    ~~~curl
+    "profile": {
+            "groupwhitelist": [
```
Do you need to indicate the attribute name you chose 'groupwhitelist' is arbitrary? The user is free to choose whatever name he wants.

mystiberry-okta replied:
done.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+
+    {% img group-claim.png %}
+    
+    In this example, the value specified in **Mapping** is `getFilteredGroups(app.profile.groupwhitelist, "group.name", 40)`.
```
Can we provide another example where one just uses a single groupId in the getFilteredGroups() EL use?

mystiberry-okta replied:
I forgot to add this. Will get it in the next draft.

mystiberry-okta replied:
Done. I added it as a comment since the only difference is the value of `valueType`.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+      -d 'grant_type=authorization_code&redirect_uri=%7B%7BredirectUri%7D%7D&code=%7B%7BauthorizationCode%7D%7D'
+      ~~~
+
+6. Decode the JWT in the response to see that the groups are in the token. 
```
What would be really helpful is to show the decoded token in the example. I would like to see how passing in "group.name" evaluates into the token I expect. Is it also worth discussing that "group.name" is an attribute on the group? Or do we do a good enough job of that in our EL documentation? If the EL documentation does a good job then might be worth saying that you can choose any attribute of the group to include into the token.

mystiberry-okta replied:
Done. Saved as text so as to be agnostic toward decoder app. I don't think this how-to topic is the right place to explain the data model.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+    curl -X POST \
+      http://{yourOktaDomain}.com/api/v1/apps \
+      -H 'accept: application/json' \
+      -H 'authorization: SSWS 00nWY8cjOh8nhyf-Un1SNZh-_ujW0bRee3jOYtT1-M' \
```
We should probably mask the session token here.

mystiberry-okta replied:
Great catch. Other doc on the site just uses ${api_token}

johngronberg-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+
+    If you want to use this group whitelist for every client that gets this claim in a token, put the `groupId` in the first parameter of the `getFilteredGroups` function described below. 
+ 
+4. Add an ID token claim on Custom Authorization Server in the Okta UI with the following function: `getFilteredGroups({group ID list}, {value-to-represent-the-group-in-the-token}, {maximum number of groups to include in token. Must be less than 100.})`.
```
+1 Going all API also means we can skip screenshots :)

mystiberry-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,132 @@
+    
+3. PUT the group IDs in the profile property bag: /{yourOktaDomain}.com/api/v1/apps/:clientId
+    
+    Request Example Fragment:
```
Not sure what you mean by "use the response from the Step x" but understand the overall concern.

mystiberry-okta commented:
Sorry, fixed a file in the wrong branch.

mystiberry-okta commented:
I haven't added the JWT yet as I haven't successfully executed step 2. @johngronberg-okta or anyone in @federations-okta , I could use a little help figuring out what I'm doing wrong with the request in step 2. Sorry!

mystiberry-okta commented:
Ready for review!

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,203 @@
+### Before You Start
+
+* Create an OAuth 2.0 client with the [Apps API](/docs/api/resources/apps.html#request-example-8). In the instruction examples, the client ID is `0oabskvc6442nkvQO0h7`.
+You can also create an OpenID Connect client, but you can also use the Okta UI to create a groups claim for OpenID Connect clients.
```
This is awkward. The first sentence indicates you can create a OAuth 2.0 client with the apps api. You then say you can also create a OIDC client and use the UI to create a group claim. I am not sure what your intension is here. I think you want to say you can create a OAuth/OIDC client via UI or apps API. I don't understand the intention for a group claim in the UI.

mystiberry-okta replied:
Yah, I'm a bit tangled up here. Not even sure what we mean by "OAuth app." It's not in the create wizard anywhere.

mystiberry-okta replied:
I cleaned this up.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,203 @@
+
+### Step One: Get the Group IDs
+
+Send a request to `/{yourOktaDomain}.com/api/v1/groups` and collect the IDs for all the groups you'll want to whitelist.
```
/{yourOktaDomain} -> {yourOktaDomain}

mystiberry-okta replied:
Fixed.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,203 @@
+### Step Two:  Add List of Groups to Profile of Client App
+ 
+If you only have one or two groups to specify, simply add the group IDs to the Okta EL function in the next step.
+However, if you have a lot of groups to whitelist, you can put the group IDs in the client app's profile property bag: `/{yourOktaDomain}.com/api/v1/apps/:aid`.
```
Same comment as above to remove leading slash

mystiberry-okta replied:
Fixed.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,203 @@
+      }
+    ~~~
+    
+You can add either app groups or user groups to the group white list property. The property value must be an array. 
```
I think we say application groups in our other documentation, not app groups.

mystiberry-okta replied:
fixed.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,203 @@
+Send a request for an ID token: 
+
+    ~~~HTTP
+    GET /oauth2/ausain6z9zIedDCxB0h7/v1/authorize?client_id=0oabskvc6442nkvQO0h7
```
This is weird to mix CURL and the GET. I would choose one and stick with it.

mystiberry-okta replied:
Fixed.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,203 @@
+    fST36rCGaDHmcm3L60JHlEeIbRArkp_793BB8OKRLdh9weNtcA
+    ~~~
+    
+    {% img group-claim-decoded.png "Group Claim in the JWT" alt:"JWT Decoder UI" %}
```
This image should be {yourOktaDomain}

mystiberry-okta replied:
Great catch!

cwu-okta replied:
@mystiberry-okta  This isn't resolved. I still see the image with {...}.oktapreview.com. Instead isn it {yourOktaDomain}.com ?

cwu-okta replied:
@mystiberry-okta  I see you did update the doc to include a textual version of the decoded token, with the correct url. I think you still need to remove _source/_assets/img/group-claim-decoded.png since it is not referenced anywhere

mystiberry-okta replied:
Done.

mystiberry-okta commented:
I added a link to this new doc in the oauth2.md, and added a summary statement at the end, as well as fixing the bugs called out. Ready for another review!

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,225 @@
+  }
+~~~
+    
+You can add either application groups or user groups to the group white list property. The property value must be an array. 
```
Should this be "white list" or "whitelist" ? I think you've been using whitelist (one word) throughout this doc.

Technically you can add a mixture of both app groups and user groups. Not sure if you want to specify that here but something to call out in case you do want to add that note.

mystiberry-okta replied:
Ah, I thought you had to have only one type in the list. I'll clarify.

cwu-okta replied:
What are your thoughts about specifying a mixture of app groups and user groups?

mystiberry-okta replied:
Since it's just an array of IDs, I don't think it's worth it to revise this topic (we say at the top "we're just using one ID to keep it simple"), add another group to the intro.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,225 @@
+
+The token contains the West Coast Division group so it can be used to evaluate against the claim defined in the authorization server.
+
+> Reminder: For flows other than implicit, post to the token endpoint `/{yourOktaDomain}.com/oauth2/:authorizationServerId/v1/token` with the user or client you want. Make sure the user is assigned to the app and to one of the groups from your whitelist.
```
Remove leading slash in url

mystiberry-okta replied:
Fixed.

yuliu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,225 @@
+	"status": "ACTIVE",
+	"claimType": "RESOURCE",
+	"valueType": "EXPRESSION",
+	"value": "\"getFilteredGroups(app.profile.groupwhitelist, "group.name", 40)\"",
```
Does the value need ""? Will the claim just be the expression itself?

mystiberry-okta replied:
@yuliu-okta, we worked on this together, and saw this in the payload (with the quotes on group.name):
 "mygroupWhitelist": [
    "WestCoastDivision"
  ],

I was wondering why it wasn't somehow labeled a claim in the payload.  But it didn't work at all without the quotes. Any suggestions?

yuliu-okta replied:
I mean the the quotes for the getFilteredGroups(), remember that it just show the expression itself getFilteredGroups(app.profile.groupwhitelist, "group.name", 40) if it has the quotes.
Also the the "group.name" should be \"group.name\"

mystiberry-okta replied:
So, the function does require quotes. It's here: https://developer.okta.com/reference/okta_expression_language/index

and also I retested.   I can't see any difference between "group.name" and group.name", what am I missing?

yuliu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,225 @@
+  
+>Hint: You can also see this value in the Okta user interface for claims, under **Mapping**: `getFilteredGroups(app.profile.groupwhitelist, "group.name", 40)`.
+
+If you had only one group to specify and so didn't need the profile, the `valueType` could simply be: `"groupID"`, in this example `00gbso71miOMjxHRW0h7`.
```
what is the valueType?

mystiberry-okta replied:
Ack, I believe it should be "the `value` could simply by the group ID, in this example...

yuliu-okta replied:
So in our [api doc]( https://developer.okta.com/reference/okta_expression_language/index#getfilteredgroups-details), we call the first parameter as whitelist, so in this case, the user should put {"00gbso71miOMjxHRW0h7"} for the whitelist

mystiberry-okta replied:
No, I'm just pointing out here that if you don't need the whitelist at all, you can supply one group ID. Maybe I should just leave this out?

yuliu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,225 @@
+}
+~~~
+
+The token contains the West Coast Division group so it can be used to evaluate against the claim defined in the authorization server.
```
What does it mean? Why the token is used to evaluate against claim?

mystiberry-okta replied:
The goal is to have a token with a claim in it that is derived from the value of the getFilteredGroups expression. I changed to this:

The token contains a claim [but it doesn't really, does it? see previous questions about group.name] for the group(s) specified in the Okta profile and configured in the authorization server.

yuliu-okta replied:
The token contains the West Coast Division group so the audience can get the group information about the user.

mystiberry-okta replied:
Ack. I put `aud` in parens after "audience" .

mystiberry-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,225 @@
+excerpt: How to create an ID token or access token that contains a groups claim
+---
+
+## How to Create a Token with a Groups Claim
```
Changed to How to Create a Token with a Groups Claim Using the Okta Profile (about as long as a title can get and not suffer SEO punishment).

yuliu-okta replied:
In UI we call it "Groups claim expression", so how about "Create a Token with a Groups Claim Using Expression"

mystiberry-okta replied:
So, the point of this article is how to use the profile to create a list of groups that you can reuse. We're not covering all the ways you can use an expression...

mystiberry-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -2308,6 +2308,8 @@ If `valueType` is `GROUPS`, then the groups returned are filtered according to t
 * `CONTAINS`: Group names contain `value` (not case sensitive). For example, if `value` is `group1`, then `MyGroup123` and `group1` are included.
 * `REGEX`: Group names match the REGEX expression in `value` (case sensitive). For example if `value` is `/^[a-z0-9_-]{3,16}$/`, then any group name that has at least 3 letters, no more than 16, and contains lower case letters, a hyphen, or numbers. 
 
+If you have many groups to specify, you can simplify the process by [creating a groups whitelist](/docs/how-to/creating-token-with-groups-claim.html).
```
Clarified: If you have many groups to specify, you can [create a groups whitelist](/docs/how-to/creating-token-with-groups-claim.html) and reuse the list in different claims.

yuliu-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -2308,6 +2308,8 @@ If `valueType` is `GROUPS`, then the groups returned are filtered according to t
 * `CONTAINS`: Group names contain `value` (not case sensitive). For example, if `value` is `group1`, then `MyGroup123` and `group1` are included.
 * `REGEX`: Group names match the REGEX expression in `value` (case sensitive). For example if `value` is `/^[a-z0-9_-]{3,16}$/`, then any group name that has at least 3 letters, no more than 16, and contains lower case letters, a hyphen, or numbers. 
 
+If you have many groups to specify, you can [create a groups whitelist](/docs/how-to/creating-token-with-groups-claim.html) and reuse the list in different claims.
```
How about "If you have complex filters of groups, you can create a groups whitelist to put them all into a claim".

mystiberry-okta replied:
Done!

yuliu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+---
+layout: docs_page
+title: How to Create a Token with a Groups Claim
```
Do we also need to change the title?

mystiberry-okta replied:
Ack.

yuliu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+excerpt: How to create an ID token or access token that contains a groups claim
+---
+
+## How to Create a Token with a Groups Claim Using the Okta Profile
```
What does the "Okta Profile" mean? is it "app.profile"?

mystiberry-okta replied:
Yes! Let me fix that in title and in the text, I have that term somewhere in the text.

yuliu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+---
+layout: docs_page
+title: Create a Token with a Groups Claim Using the App Profile
+excerpt: How to use the app profile to create an ID token or  access token that contains a groups claim
```
extra space "or  access token"

mystiberry-okta replied:
ack

mystiberry-okta replied:
ack

yuliu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+
+### Step Two:  Add List of Groups to Profile of Client App
+ 
+If you only have one or two groups to specify, simply add the group IDs to the Okta EL function in the next step.
```
simply add the group IDs to the first parameter of the `getFilteredGroups` function described in the next step

mystiberry-okta replied:
ack

mystiberry-okta replied:
ack

larsjohansen-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+
+## How to Create a Token with a Groups Claim Using the App Profile
+
+You can add a groups claim for any application group or user group into ID tokens or access tokens for API Access Management.
```
```application group or user group``` -> ```combination of application groups and user groups```

mystiberry-okta replied:
ack

larsjohansen-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+* Create an OAuth 2.0 or OpenID Connect client with the [Apps API](/docs/api/resources/apps.html#request-example-8). In the instruction examples, the client ID is `0oabskvc6442nkvQO0h7`.
+* Create the groups that you wish to configure in the groups claim. In the instruction examples, we're configuring the `WestCoastDivision` group, and the group ID is `00gbso71miOMjxHRW0h7`.
+
+To create a a groups claim, assign a group whitelist to your client app, and configure a groups claim that references the whitelist.
```
```a a groups``` -> ```a groups```

mystiberry-okta replied:
ack

jmelberg-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+
+### Step One: Get the Group IDs
+
+Send a request to `{yourOktaDomain}.com/api/v1/groups` and collect the IDs for all the groups you'll want to whitelist.
```
`{yourOktaDomain}.com/api/v1/groups` -> `https://{yourOktaDomain}.com/api/v1/groups`

mystiberry-okta replied:
ack

jmelberg-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+~~~json
+ [
+    {
+            "id": "00gbso71miOMjxHRW0h7",
```
Looks like we have `8` spaces here instead of `4`

mystiberry-okta replied:
ack

jmelberg-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+### Step Two:  Add List of Groups to Profile of Client App
+ 
+If you only have one or two groups to specify, simply add the group IDs to the first parameter of the `getFilteredGroups` function described in the next step.
+However, if you have a lot of groups to whitelist, you can put the group IDs in the client app's profile property bag: `{yourOktaDomain}.com/api/v1/apps/:aid`.
```
`{yourOktaDomain}.com/api/v1/apps/:aid` -> `https://{yourOktaDomain}.com/api/v1/apps/:aid`

mystiberry-okta replied:
ack (in future, you can mark it once and say "all of these" so you don't waste your time.)

jmelberg-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+
+### Step Four: Test that the Claim Is Delivered in the Token
+
+Send a request for an ID token to `{yourOktaDomain}.com/oauth2/:authorizationServerId/v1/authorize` to obtain a token with the right groups claim.
```
`{yourOktaDomain}.com/oauth2/:authorizationServerId/v1/authorize` -> `https://{yourOktaDomain}.com/oauth2/:authorizationServerId/v1/authorize`

mystiberry-okta replied:
ack

jmelberg-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+
+The token contains the group WestCoastDivision so the audience (`aud`) has access to the group information about the user.
+
+> Reminder: For flows other than implicit, post to the token endpoint `{yourOktaDomain}.com/oauth2/:authorizationServerId/v1/token` with the user or client you want. Make sure the user is assigned to the app and to one of the groups from your whitelist.
```
Prepend `https:`

mystiberry-okta replied:
Ack

larsjohansen-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+  }
+~~~
+    
+You can add application groups, user groups or both to the group whitelist property. The property value must be an array. 
```
whitelist or groupwhitelist property? We should be consistent with the example.

mystiberry-okta replied:
I'm revising all of them to be group whitelist, except when repetition of the word "group" 
 in a single sentence would make the sentence hard to read.

larsjohansen-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+    
+You can add application groups, user groups or both to the group whitelist property. The property value must be an array. 
+
+To use this group whitelist for every client that gets this claim in a token, put the group whitelist in the first parameter of the `getFilteredGroups` function described below. 
```
```To use this group whitelist``` sounds wrong, as it's not this list, it's the list for each client using the same json structure to reference it. Perhaps: ```To use the group whitelist for every client that gets this claim in a token, put the path to the whitelist in the first parameter of the `getFilteredGroups` function described below. ```

mystiberry-okta replied:
ack

larsjohansen-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+ 
+### Step Three: Configure a Custom Claim for Your Groups
+
+Add a custom claim for the ID token on a Custom Authorization Server with the following function: `getFilteredGroups({app.profile.whitelist-name}, "{value-to-represent-the-group-in-the-token}", {maximum number of groups to include in token. Must be less than 100.})`.
```
```whitelist-name``` is very confusing as we already have ```groupwhitelist``` as an example and also referenced this as simply ```whitelist```. Let's try to make this consistent across.

mystiberry-okta replied:
I added -name to make it more clear that it's a variable, not literal (can't use italics inside the 'ticks'). Since this is also confusing, will remove. Plus I just figured out how to do italics inside code.

mystiberry-okta replied:
ah, it doesn't work. But the example is right there, and we mention the name throughout, so hoping this won't confuse folks.

larsjohansen-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+  -H 'cache-control: no-cache' \
+  -H 'content-type: application/json' \
+  -d '{
+	"name": "mygroupWhitelist",
```
hmm, what about just simply "groups"? This claim should have nothing to do with the whitelist in the el function its using.

mystiberry-okta replied:
I was worried that it wouldn't be clear what was going on. Will change to groups.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+You can add a groups claim for any application group or user group into ID tokens or access tokens for API Access Management.
+This process optionally uses Okta's flexible app profile, which accepts any JSON-compliant content, to create a whitelist of groups
+that can then easily be referenced. This is especially useful if you have a large number of groups to whitelist or otherwise
+need to dynamically set group whitelists on a per-application basis.
```
@larsjohansen-okta  Is this technically accurate? We don't dynamically set the groups. Technically it is very static. The admin configures a static list of groups on the client's profile. In the token definition he uses an EL expression to indicate which attribute to include for each of the groups in the static list. Is the terminology dynamic correct here?

@mystiberry-okta

larsjohansen-okta replied:
@cwu-okta your understanding is correct, and I agree this sounds funky; setting the group whitelist on the app profile is a static operation so we shouldn't use the word dynamically set group whitelist there. However what _is_ dynamically determined upon evaluation is which of those whitelisted groups will appear in the token and that depends on which of the groups the user is member of.

cwu-okta replied:
Fair enough. I think I am particularly sensitive to the word dynamic since we've had some customers indicate the current mechanism isn't truly dynamic and they've requested a feature enhancement to make it really dynamic.

@mystiberry-okta  Can we reword, taking into account Lars' comment above?

mystiberry-okta replied:
I will remove 'dynamic' here and beef up explanation around what happens--not sure it's clear that there might be 20 groups in whitelist, but only those groups that the user is a member of appear in the token.

mystiberry-okta replied:
@larsjohansen-okta , I added this: Once you create the group whitelist and add it to an app, tokens for that app will contain the groups from the whitelist if the user is a member.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+ 
+### Step Three: Configure a Custom Claim for Your Groups
+
+Add a custom claim for the ID token on a Custom Authorization Server with the following function: `getFilteredGroups({app.profile.whitelist-name}, "{value-to-represent-the-group-in-the-token}", {maximum number of groups to include in token. Must be less than 100.})`.
```
Since this is a guide, do we want to mention that you can either use the API or the UI to set this value? Should we provide instructions for how to accomplish this via UI too?

mystiberry-okta replied:
This draft started with UI instructions for this step. I don't think it's helpful for an instruction to list multiple ways of completing a step, it typically makes the task doc harder to read, the instructions harder to follow. We should pick the best way, and then use that way. Fine to mention the other way and point them to starting point in the UI in a note. Are there any other steps that should also have that note? Or a simple reminder at the top of the topic that some steps can be done in the UI as well?

mystiberry-okta replied:
I added this note to the top: > Note: You can perform Step One and Step Three in the user interface if you prefer.  and modified the sentence about "you can see" to "you can configure" and gave more specific path (since it's not clear from context).

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+  "idp": "00o5t60il3UzyIe5v0h7",
+  "nonce": "2185edf4-84a8-491e-ba72-20b765402543",
+  "auth_time": 1505320024,
+  "mygroupWhitelist": [
```
We might need to add more explanation about which mapping was added to add this value into the token. This attribute in 'mygroupWhitelist' is not mentioned anywhere within this guide.

mystiberry-okta replied:
Not sure what you mean. WestCoastDivision is the thing we've been trying to add throughout the whole task, and is mentioned twice before here and once after.

mystiberry-okta replied:
Ack.

mystiberry-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,223 @@
+    "WestCoastDivision"
+  ],
+  "firstName": "Annie"
+}
```
@cwu-okta or @larsjohansen-okta should I mention "Annie" at the start of this topic as the user being evaluated? In the real request it's me of course--not sure how the request in step 4 picked me. Because I sent the request (what we mean by user context?)? Do I need to explain this here?

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,229 @@
+that can then easily be referenced. This is especially useful if you have a large number of groups to whitelist or otherwise
+need to set group whitelists on a per-application basis.
+
+Once you create the group whitelist and add it to an app, tokens for that app will contain the groups from the whitelist if the user is a member. 
```
This sounds slightly dangling and also takes a few presumptions, ie that you set up the token with the configuration to use the group. Is this document strictly focused for how to add groups for API AM? If it is then we should probably reword to be more explicit.

Suggestion which you can feel free to iterate on.
```Once a group whitelist is configured in the OAuth 2.0 client, configure the authorization server to reference the group white list. An ID or refresh token minted by the authorization server will contain the whitelisted groups which the user is a member of.``` (yes I know there is a dangling participle here. I am hoping you can use some grammar-foo to tell me how to fix it.)

mystiberry-okta replied:
Yes it is strictly focused on API AM because it requires a custom AS. We can shelve this until John returns and verify that was his intention (origianlly I thought it wasn't).  Please advise.

cwu-okta replied:
Great, so if it is focused on API Access Management, then can we alter the sentence, using my suggestion as a starting point?

mystiberry-okta replied:
Ack, per discussion.

mystiberry-okta replied:
@cwu-okta , I took a stab at making it easier to read, simpler. What do you think? 

Once you complete Steps 1-3, tokens for an app will contain the groups from the whitelist as shown in [Step Four](#step-four-test-that-the-claim-is-delivered-in-the-token).
Remember that the groups from the whitelist are added to the token only if the user is a member of the group.

cwu-okta replied:
I think mentioning the steps in the introduction are a bit premature. I would rather include the note that if a user is a member of the group, that group is included in the token where it is relevant, ie in the step that describes evaluating the EL expression and including the matching groups into the token.

mystiberry-okta replied:
Ack. Removed both sentences.

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,229 @@
+
+See [group function documentation](/reference/okta_expression_language/#group-functions) for more information about specifying groups with `getFilteredGroups`.
+
+In the token, the dynamically evaluated list of groups appears under the `name` you specify. In this example, `groups`.
```
What is "In this example, 'groups'"? What is "groups"?

mystiberry-okta replied:
The sentence before it explains.

cwu-okta replied:
I don't think I see it. Can you point it out explicitly?

I think the intention was the 'groups' attribute is to be used for this, but it appears out of place since I don't see it referenced anywhere in this doc up until this point. The only reference I see it is in the actual example further down, line 220.

mystiberry-okta replied:
How about "in Step 4 Payload Data,".   This element was named mygroupswhitelist before, and someone else said that was too confusing. This has been here forever, so I'm wondering if the explnaation is actually making it more confusing?

mystiberry-okta replied:
How about this: 
1. Put this sentence right before the example in step 3: 
In this example, the `name` for the claim is `groups`, but you can name it whatever you wish.

2.  Stce from line 170: Now when you mint a token, groups in the `groupwhitelist` that also have the user as a member are included in the `groups` claim. Test your configuration in the next step.

cwu-okta replied:
sounds good

cwu-okta commented on _source/_docs/how-to/creating-token-with-groups-claim.md:
```diff
@@ -0,0 +1,229 @@
+    
+You can add application groups, user groups or both to the group whitelist, specified as an array of IDs. 
+
+To use the group whitelist for every client that gets this claim in a token, put the path to the whitelist in the first parameter of the `getFilteredGroups` function described below. 
```
Nit: ```put the path to the whitelist``` sounds funny here since it is correct but no one thinks of it that way. Instead, it is the attribute name. Would be better to alter this to be ```put the attribute name of the whitelist...```

mystiberry-okta replied:
Ack, will change and merge.

mystiberry-okta commented:
Email  approval from cwu-okta
